src/Microsoft.TestPlatform.CrossPlatEngine/DataCollection/InProcDataCollector.cs

```diff
@@ -67,10 +68,20 @@ internal InProcDataCollector(string codeBase, string assemblyQualifiedName, Type
             this.assemblyLoadContext = assemblyLoadContext;
 
             var assembly = this.LoadInProcDataCollectorExtension(codeBase);
-            this.dataCollectorType =
-                assembly?.GetTypes()
-                    .FirstOrDefault(x => x.AssemblyQualifiedName.Equals(assemblyQualifiedName) && interfaceTypeInfo.IsAssignableFrom(x.GetTypeInfo()));
 
+            Func<Type, bool> filterPredicate;
+            if (Path.GetFileName(codeBase) == Constants.CoverletDataCollectorCodebase)
+            {
+                // If we're loading coverlet collector we skip to check the version of assembly
+                // to allow upgrade throught nuget package
+                filterPredicate = (x) => x.FullName.Equals(Constants.CoverletDataCollectorTypeName) && interfaceTypeInfo.IsAssignableFrom(x.GetTypeInfo());
```

Comment:
@mayankbansal018 Should we make this default ?

Reply:
I think no, If I'm not mistaken that aqn is read from possible runsetting file so a user should have that assembly in local folder. The issue with coverlet is that we(you) inject hardcoded aqn if user provide `--collect:"XPlat Code Coverage"`.
It works well for user provided collectors because there is always a runsettings file provided, but with coverlet don't, so we need a way to make it works for every version without runsettings file.
So my idea is to check only "trusted" type/lib name only for coverlet.
BTW I could miss something in the big picture.